### DISCUSSION TITLE
FastAPI: Some requests are failing due to 10s timeout

### DISCUSSION THREAD

**Discussion by rileyhun** [fastapi/fastapi#8696] [Questions] [answered]
### First Check

- [X] I added a very descriptive title to this issue.
- [X] I used the GitHub search to find a similar issue and didn't find it.
- [X] I searched the FastAPI documentation, with the integrated search.
- [X] I already searched in Google "How to X in FastAPI" and didn't find any information.
- [X] I already read and followed all the tutorial in the docs and didn't find an answer.
- [X] I already checked if it is not related to FastAPI but to [Pydantic](https://github.com/samuelcolvin/pydantic).
- [X] I already checked if it is not related to FastAPI but to [Swagger UI](https://github.com/swagger-api/swagger-ui).
- [X] I already checked if it is not related to FastAPI but to [ReDoc](https://github.com/Redocly/redoc).

### Commit to Help

- [X] I commit to help with one of those options 👆

### Example Code

The gunicorn configuration we are using:
```python
"""gunicorn server configuration."""
import os
​
threads = 2
workers = 4
timeout = 60
keepalive = 1800
graceful_timeout = 1800
bind = f":{os.environ.get('PORT', '80')}"
worker_class = "uvicorn.workers.UvicornWorker"
```

### Description

We have deployed a model prediction service in production that is using FastAPI and unfortunately, some of the requests are failing due to a 10s timeout. In terms of concurrent requests, we typically only load about 2/3 requests per second, so I wouldn't think that would be too much strain on FastAPI. The first thing we tried to do is isolate the FastAPI framework from the model itself, and when we performed some tracing, we noticed that a lot of time (6 seconds) was spent on this segment: `starlette.exceptions:ExceptionMiddleware.__call__`.

Would really appreciate some guidance on what the above segment implies and what is causing timeout issues for some requests under a not too strenuous load.

<img width="1176" alt="Screen Shot 2021-11-30 at 12 02 21 PM" src="https://user-images.githubusercontent.com/20649466/144510696-f8901dae-43b2-49cc-a916-37fa7e06eb1b.png">

 

### Wanted Solution

I would like to optimize my FastAPI application so that latency for all requests are consistent and there are no outliers that error out due to the 10s timeout. 

### Wanted Code

```python
N/A
```

### Operating System

macOS

### FastAPI Version

0.65.2

### Python Version

3.7.10

**Comment by Dustyposa** [comment]
How many  concurrent requests when you test?
I guess the thread is full, new request need to wait new thread when release a thread.

**Comment by rileyhun** [comment]
Just 3 per second probably. We've been monitoring things and CPU usage is ok. I don't think it's any issue in terms of threads either. It's some application-level issue related to the `middle-ware`, perhaps?

**Comment by jgould22** [comment]
Are you blocking the event loop by doing a substantial CPU bound task on the event loop ?

**Comment by rileyhun** [comment]
It's a Bert Model so yes, but there's middleware that sends metrics to Prometheus that seems to be taking a long time as well

**Comment by jgould22** [comment, marked as answer]
You need to do your heavy processing outside of the event loop.

Async Python is Cooperative Multitasking, if one of your requests gets a hold of the event loop and doesnt call an await for a long period of time the other async tasks on the event loop will not make any progress (in your prometheus code for example).

Offload your CPU bound tasks to another process (people often use Celery ) so that you are not blocking the loop.

I think you could check this by changing your controller code to just a def so FastAPI will spawn it on a thread pool as described here https://fastapi.tiangolo.com/async/ 

If you are already doing then than I am not sure what the issue might be.
